File: Problem_2.java
// Time Complexity : O(m*n)
// Space Complexity :O(m*n)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// if not * then take it from the horizontal top, if is * then zero one case for zero: take it from 2 step back, for one take or between old value and one above if both the character is same
// Your code here along with comments explaining your approach
class Solution {
    public boolean isMatch(String s, String p) {
        int sl = s.length(); int pl = p.length();
        if(s.equals(p)) return true;
        boolean [][] dp = new boolean [sl+1][pl+1];
        dp[0][0] = true;
        // top row
        for(int j = 1; j < dp[0].length; j++){
            if(p.charAt(j-1) == '*'){
                dp[0][j] = dp[0][j-2];
            }
        }
        // main
        for( int i = 1; i < dp.length; i++){
            for( int j = 1; j < dp[0].length; j++){
                if(p.charAt(j-1) != '*'){
                    // if char or '.'
                    if(p.charAt(j-1) == s.charAt(i-1) || p.charAt(j-1) == '.'){
                        dp[i][j] = dp[i-1][j-1];
                    }
                }else{
                    // zero case
                    dp[i][j] = dp[i][j-2];
                    // one case
                    if(p.charAt(j-2) == s.charAt(i-1) || p.charAt(j-2) == '.'){
                        dp[i][j] = dp[i][j] || dp[i-1][j];
                    }
                }
            }
        }
        return dp[sl][pl];
    }
}

File: Problem2.py
#Time complexity is: O(mn)
#Space complexity is O(mn)
#No issues faced while coding
#Code ran successfully on leetcode
class Solution(object):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        #Base condition
        if s == p:
            return True
        #Initializing all the required variables
        m = len(s)
        n = len(p)
        dp = [[False for _ in range(n+1)] for _ in range(m+1)]
        #Intializing the value to True in top left corner
        dp[0][0] = True
        #Moving over first row in the matrix
        for j in range(1, n+1):
            c = p[j-1]
            #If the current character is *, we will replace that 
            #with the value two stpes back
            if c == "*":
                dp[0][j] = dp[0][j-2]
        #Iterating through the matrix
        for i in range(1, m+1):
            for j in range(1, n+1):
                #Taking the required characters in the s and p string
                sChar = s[i-1]
                pChar = p[j-1]
                #If the incoming characters in the string and pattern matches
                #or character in the pattern is equal to ".", we will replace that
                #To the diagonal top left corner
                if sChar == pChar or pChar == ".":
                    dp[i][j] = dp[i-1][j-1]
                #If the character is equal to *
                elif pChar == "*":
                    #Zero case
                    dp[i][j] = dp[i][j-2]
                    #Once case, where character in s is equal to p[j-2]
                    #or p[j-2]=="."
                    if sChar == p[j-2] or p[j-2] == ".":
                        #Then we will take the or value at the position dp[i][j]
                        dp[i][j] = dp[i][j] or dp[i-1][j]
        #Finally we are returning dp[m][n] value
        return dp[m][n]
